state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1193,106 +1506,139 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
-				} => DepositAction::Swap {
-					swap_id: T::SwapDepositHandler::schedule_swap_from_channel(
-						<<T::TargetChain as Chain>::ChainAccount as IntoForeignChainAddress<
-							T::TargetChain,
-						>>::into_foreign_chain_address(deposit_address.clone()),
-						block_height.into(),
-						asset.into(),
-						destination_asset,
-						amount_after_fees.into(),
-						destination_address,
-						broker_id,
-						broker_commission_bps,
-						channel_id,
-					),
+				Some(_) => {
+					// Boosted deposit is found but the amounts didn't match, the deposit
+					// should be processed as not boosted.
+					None
 				},
-				ChannelAction::CcmTransfer {
-					destination_asset,
-					destination_address,
-					channel_metadata,
-					..
-				} => {
-					if let Ok(CcmSwapIds { principal_swap_id, gas_swap_id }) =
-						T::CcmHandler::on_ccm_deposit(
-							asset.into(),
-							amount_after_fees.into(),
-							destination_asset,
-							destination_address,
-							CcmDepositMetadata {
-								source_chain: asset.into(),
-								source_address: None,
-								channel_metadata,
-							},
-							SwapOrigin::DepositChannel {
-								deposit_address: T::AddressConverter::to_encoded_address(
-									<T::TargetChain as Chain>::ChainAccount::into_foreign_chain_address(
-										deposit_address.clone(),
-									),
-								),
-								channel_id,
-								deposit_block_height: block_height.into(),
-							},
-						) {
-						DepositAction::CcmTransfer { principal_swap_id, gas_swap_id }
-					} else {
-						DepositAction::NoAction
-					}
+				None => {
+					log_or_panic!("Could not find deposit by boost id: {boost_id}");
+					// This is unexpected since we always add a prewitnessed deposit at the
+					// same time as boosting it! Because we won't be able to confirm if the
+					// amount is correct, we fallback to processing the deposit as not boosted.
+					None
 				},
-			};
+			}
+		} else {
+			// The channel is not even boosted, so we process the deposit as not boosted
+			None
+		};
+
+		if let Some((boost_id, used_pools)) = maybe_boost_to_process {
+			PrewitnessedDeposits::<T, I>::remove(channel_id, boost_id);
+
+			// Note that ingress fee is not payed here, as it has already been payed at the time
+			// of boosting
+			DepositBalances::<T, I>::mutate(asset, |deposits| {
+				deposits.register_deposit(deposit_amount)
+			});
+
+			for fee_tier in used_pools {
+				BoostPools::<T, I>::mutate(asset, fee_tier, |maybe_pool| {
+					if let Some(pool) = maybe_pool {
+						for (booster_id, amount) in pool.on_finalised_deposit(boost_id) {
+							if let Err(err) = T::LpBalance::try_credit_account(
+								&booster_id,
+								asset.into(),
+								amount.into(),
+							) {
+								log_or_panic!(
+									"Failed to credit booster account {} after unlock of {amount:?} {asset:?}: {:?}",
+									booster_id, err
+								);
+							}
+						}
+					}
+				});
+			}
+
+			// This allows the channel to be boosted again:
```

Comment:
(1) and (2) are tested in `witnessed_amount_does_not_match_boosted` deals with multiple deposits into the same channel, and as one of the things it checks is that we can't have simultaneous boosts, and that after finalisation we can boost again.
(3) is a little harder since the test are currently with EVM assets and don't get recycled, and doesn't look trivial to re-use the same channel. I see that this is something that we might want to test for completeness, but at the same time this doesn't really correspond to any new code. Re-using a channel is no different from using a brand new channel from boost's perspective (so it is actually not trivial to make this *not* work as expected). So I would say lets create a linear issue for this in the interest of time, and we can come back to this later.

Reply:
Created this: https://linear.app/chainflip/issue/PRO-1285/test-that-channel-can-be-boosted-after-recycling